File: sdk-typescript/packages/collab/src/collab.ts
/**
 * WAVE SDK - Collab API
 *
 * Real-time collaboration features for projects and media.
 *
 * NOTE: This is a client SDK. All authorization checks are performed server-side.
 * The API will return 403 Forbidden if the user lacks required permissions.
 */

import type {
  WaveClient,
  PaginationParams,
  PaginatedResponse,
} from '@wave-av/core';

export * from './collab-types';
import type {
  Annotation,
  CollabRoom,
  Comment,
  CreateRoomRequest,
  InviteRequest,
  ListRoomsParams,
  Participant,
  ParticipantPermissions,
  ParticipantRole,
  UpdateRoomRequest,
} from './collab-types';
export type { RoomStatus, ParticipantRole, PresenceStatus, CollabRoom, RoomSettings, Participant, ParticipantPermissions, CursorPosition, Selection, Comment, Reaction, Annotation, CreateRoomRequest, UpdateRoomRequest, InviteRequest, ListRoomsParams } from './collab-types';

export class CollabAPI {
  private readonly client: WaveClient;
  private readonly basePath = '/v1/collab';

  constructor(client: WaveClient) {
    this.client = client;
  }

  // Rooms

  /**
   * Create a collaboration room
   *
   * Requires: collab:create permission
   */
  async createRoom(request: CreateRoomRequest): Promise<CollabRoom> {
    return this.client.post<CollabRoom>(`${this.basePath}/rooms`, request);
  }

  /**
   * Get a room by ID
   *
   * Requires: collab:read permission
   */
  async getRoom(roomId: string): Promise<CollabRoom> {
    return this.client.get<CollabRoom>(`${this.basePath}/rooms/${roomId}`);
  }

  /**
   * Update a room
   *
   * Requires: collab:update permission
   */
  async updateRoom(roomId: string, request: UpdateRoomRequest): Promise<CollabRoom> {
    return this.client.patch<CollabRoom>(
      `${this.basePath}/rooms/${roomId}`,
      request
    );
  }

  /**
   * Close a room
   *
   * Requires: collab:manage permission
   */
  async closeRoom(roomId: string): Promise<CollabRoom> {
    return this.client.post<CollabRoom>(`${this.basePath}/rooms/${roomId}/close`);
  }

  /**
   * Archive a room
   *
   * Requires: collab:manage permission (server-side RBAC enforced)
   */
  async archiveRoom(roomId: string): Promise<void> {
    await this.client.delete(
      `${this.basePath}/rooms/${roomId}`,
      { method: 'DELETE' }
    );
  }

  /**
   * List rooms
   *
   * Requires: collab:read permission
   */
  async listRooms(params?: ListRoomsParams): Promise<PaginatedResponse<CollabRoom>> {
    return this.client.get<PaginatedResponse<CollabRoom>>(
      `${this.basePath}/rooms`,
      { params: params as Record<string, string | number | boolean | undefined> }
    );
  }

  /**
   * Get join token for real-time connection
   *
   * Requires: collab:join permission
   */
  async getJoinToken(
    roomId: string,
    options?: { display_name?: string; avatar_url?: string }
  ): Promise<{ token: string; expires_at: string; websocket_url: string }> {
    return this.client.post(`${this.basePath}/rooms/${roomId}/token`, options);
  }

  // Participants

  /**
   * List participants in a room
   *
   * Requires: collab:read permission
   */
  async listParticipants(
    roomId: string,
    params?: PaginationParams
  ): Promise<PaginatedResponse<Participant>> {
    return this.client.get<PaginatedResponse<Participant>>(
      `${this.basePath}/rooms/${roomId}/participants`,
      { params: params as Record<string, string | number | boolean | undefined> }
    );
  }

  /**
   * Get a participant
   *
   * Requires: collab:read permission
   */
  async getParticipant(roomId: string, participantId: string): Promise<Participant> {
    return this.client.get<Participant>(
      `${this.basePath}/rooms/${roomId}/participants/${participantId}`
    );
  }

  /**
   * Update a participant's role
   *
   * Requires: collab:manage permission
   */
  async updateParticipant(
    roomId: string,
    participantId: string,
    updates: { role?: ParticipantRole; permissions?: Partial<ParticipantPermissions> }
  ): Promise<Participant> {
    return this.client.patch<Participant>(
      `${this.basePath}/rooms/${roomId}/participants/${participantId}`,
      updates
    );
  }

  /**
   * Remove a participant from a room
   *
   * Requires: collab:manage permission (server-side RBAC enforced)
   */
  async removeParticipant(roomId: string, participantId: string): Promise<void> {
    await this.client.delete(
      `${this.basePath}/rooms/${roomId}/participants/${participantId}`,
      { method: 'DELETE' }
    );
  }

  /**
   * Invite users to a room
   *
   * Requires: collab:invite permission
   */
  async invite(
    roomId: string,
    invites: InviteRequest[]
  ): Promise<{ sent: number; failed: Array<{ email?: string; user_id?: string; error: string }> }> {
    return this.client.post(`${this.basePath}/rooms/${roomId}/invite`, { invites });
  }

  // Comments

  /**
   * Add a comment
   *
   * Requires: collab:comment permission
   */
  async addComment(
    roomId: string,
    comment: {
      content: string;
      timestamp?: number;
      element_id?: string;
      position?: { x: number; y: number };
      parent_id?: string;
    }
  ): Promise<Comment> {
    return this.client.post<Comment>(
      `${this.basePath}/rooms/${roomId}/comments`,
      comment
    );
  }

  /**
   * List comments
   *
   * Requires: collab:read permission
   */
  async listComments(
    roomId: string,
    params?: PaginationParams & {
      resolved?: boolean;
      element_id?: string;
      parent_id?: string;
    }
  ): Promise<PaginatedResponse<Comment>> {
    return this.client.get<PaginatedResponse<Comment>>(
      `${this.basePath}/rooms/${roomId}/comments`,
      { params: params as Record<string, string | number | boolean | undefined> }
    );
  }

  /**
   * Update a comment
   *
   * Requires: collab:comment permission (own comments) or collab:manage
   */
  async updateComment(
    roomId: string,
    commentId: string,
    updates: { content?: string; resolved?: boolean }
  ): Promise<Comment> {
    return this.client.patch<Comment>(
      `${this.basePath}/rooms/${roomId}/comments/${commentId}`,
      updates
    );
  }

  /**
   * Remove a comment
   *
   * Requires: collab:comment permission (own) or collab:manage (server-side RBAC enforced)
   */
  async removeComment(roomId: string, commentId: string): Promise<void> {
    await this.client.delete(
      `${this.basePath}/rooms/${roomId}/comments/${commentId}`,
      { method: 'DELETE' }
    );
  }

  /**
   * Add a reaction to a comment
   *
   * Requires: collab:comment permission
   */
  async addReaction(roomId: string, commentId: string, emoji: string): Promise<Comment> {
    return this.client.post<Comment>(
      `${this.basePath}/rooms/${roomId}/comments/${commentId}/reactions`,
      { emoji }
    );
  }

  /**
   * Remove a reaction from a comment
   *
   * Requires: collab:comment permission (server-side RBAC enforced)
   */
  async removeReaction(
    roomId: string,
    commentId: string,
    emoji: string
  ): Promise<void> {
    await this.client.delete(
      `${this.basePath}/rooms/${roomId}/comments/${commentId}/reactions`,
      { method: 'DELETE', params: { emoji } as unknown as Record<string, string | number | boolean | undefined> }
    );
  }

  // Annotations

  /**
   * Add an annotation
   *
   * Requires: collab:annotate permission
   */
  async addAnnotation(
    roomId: string,
    annotation: Omit<Annotation, 'id' | 'room_id' | 'user_id' | 'created_at' | 'updated_at'>
  ): Promise<Annotation> {
    return this.client.post<Annotation>(
      `${this.basePath}/rooms/${roomId}/annotations`,
      annotation
    );
  }

  /**
   * List annotations
   *
   * Requires: collab:read permission
   */
  async listAnnotations(
    roomId: string,
    params?: PaginationParams & { timestamp?: number; user_id?: string }
  ): Promise<PaginatedResponse<Annotation>> {
    return this.client.get<PaginatedResponse<Annotation>>(
      `${this.basePath}/rooms/${roomId}/annotations`,
      { params: params as Record<string, string | number | boolean | undefined> }
    );
  }

  /**
   * Update an annotation
   *
   * Requires: collab:annotate permission (own) or collab:manage
   */
  async updateAnnotation(
    roomId: string,
    annotationId: string,
    updates: Partial<Omit<Annotation, 'id' | 'room_id' | 'user_id' | 'created_at' | 'updated_at'>>
  ): Promise<Annotation> {
    return this.client.patch<Annotation>(
      `${this.basePath}/rooms/${roomId}/annotations/${annotationId}`,
      updates
    );
  }

  /**
   * Remove an annotation
   *
   * Requires: collab:annotate permission (own) or collab:manage (server-side RBAC enforced)
   */
  async removeAnnotation(roomId: string, annotationId: string): Promise<void> {
    await this.client.delete(
      `${this.basePath}/rooms/${roomId}/annotations/${annotationId}`,
      { method: 'DELETE' }
    );
  }

  /**
   * Clear all annotations
   *
   * Requires: collab:manage permission
   */
  async clearAnnotations(roomId: string): Promise<{ cleared: number }> {
    return this.client.post(`${this.basePath}/rooms/${roomId}/annotations/clear`);
  }

  // Recording

  /**
   * Start recording the collaboration session
   *
   * Requires: collab:record permission
   */
  async startRecording(
    roomId: string
  ): Promise<{ recording_id: string; started_at: string }> {
    return this.client.post(`${this.basePath}/rooms/${roomId}/recording/start`);
  }

  /**
   * Stop recording
   *
   * Requires: collab:record permission
   */
  async stopRecording(
    roomId: string
  ): Promise<{ recording_id: string; url: string; duration: number }> {
    return this.client.post(`${this.basePath}/rooms/${roomId}/recording/stop`);
  }

  /**
   * Get recording status
   *
   * Requires: collab:read permission
   */
  async getRecordingStatus(
    roomId: string
  ): Promise<{ recording: boolean; recording_id?: string; started_at?: string; duration?: number }> {
    return this.client.get(`${this.basePath}/rooms/${roomId}/recording`);
  }
}

/**
 * Create a Collab API instance
 */
export function createCollabAPI(client: WaveClient): CollabAPI {
  return new CollabAPI(client);
}

File: sdk-python/wave_av_sdk/collab.py
"""WAVE SDK - Collab API. Real-time collaboration rooms, participants, comments, and annotations."""
from __future__ import annotations
from typing import Any
from pydantic import BaseModel
from wave.client import WaveClient

class CollabRoom(BaseModel):
    id: str; organization_id: str; name: str; status: str; participant_count: int = 0; max_participants: int = 50; settings: dict | None = None; created_at: str; updated_at: str

class Participant(BaseModel):
    id: str; user_id: str; name: str; role: str; status: str; permissions: dict | None = None; joined_at: str

class Comment(BaseModel):
    id: str; room_id: str; user_id: str; text: str; timestamp: float | None = None; parent_id: str | None = None; reactions: list[dict] | None = None; created_at: str; updated_at: str

class Annotation(BaseModel):
    id: str; room_id: str; user_id: str; type: str; data: dict; timestamp: float | None = None; created_at: str

class CollabAPI:
    def __init__(self, client: WaveClient): self._client = client; self._base = "/v1/collab"
    def create_room(self, name: str, **kwargs: Any) -> CollabRoom: return CollabRoom(**self._client.post(f"{self._base}/rooms", json={"name": name, **kwargs}))
    def get_room(self, room_id: str) -> CollabRoom: return CollabRoom(**self._client.get(f"{self._base}/rooms/{room_id}"))
    def update_room(self, room_id: str, **kwargs: Any) -> CollabRoom: return CollabRoom(**self._client.patch(f"{self._base}/rooms/{room_id}", json=kwargs))
    def list_rooms(self, **params: Any) -> dict: return self._client.get(f"{self._base}/rooms", params={k: v for k, v in params.items() if v is not None})
    def close_room(self, room_id: str) -> None: self._client.post(f"{self._base}/rooms/{room_id}/close")
    def archive_room(self, room_id: str) -> None: self._client.post(f"{self._base}/rooms/{room_id}/archive")
    def get_join_token(self, room_id: str) -> dict: return self._client.post(f"{self._base}/rooms/{room_id}/join-token")
    def invite(self, room_id: str, user_ids: list[str], role: str = "viewer") -> None: self._client.post(f"{self._base}/rooms/{room_id}/invite", json={"user_ids": user_ids, "role": role})
    def list_participants(self, room_id: str) -> list[Participant]: return [Participant(**p) for p in self._client.get(f"{self._base}/rooms/{room_id}/participants")]
    def get_participant(self, room_id: str, participant_id: str) -> Participant: return Participant(**self._client.get(f"{self._base}/rooms/{room_id}/participants/{participant_id}"))
    def update_participant(self, room_id: str, participant_id: str, **kwargs: Any) -> Participant: return Participant(**self._client.patch(f"{self._base}/rooms/{room_id}/participants/{participant_id}", json=kwargs))
    def remove_participant(self, room_id: str, participant_id: str) -> None: self._client.delete(f"{self._base}/rooms/{room_id}/participants/{participant_id}")
    def list_comments(self, room_id: str, **params: Any) -> list[Comment]: return [Comment(**c) for c in self._client.get(f"{self._base}/rooms/{room_id}/comments", params={k: v for k, v in params.items() if v is not None})]
    def add_comment(self, room_id: str, text: str, **kwargs: Any) -> Comment: return Comment(**self._client.post(f"{self._base}/rooms/{room_id}/comments", json={"text": text, **kwargs}))
    def update_comment(self, room_id: str, comment_id: str, text: str) -> Comment: return Comment(**self._client.patch(f"{self._base}/rooms/{room_id}/comments/{comment_id}", json={"text": text}))
    def remove_comment(self, room_id: str, comment_id: str) -> None: self._client.delete(f"{self._base}/rooms/{room_id}/comments/{comment_id}")
    def add_reaction(self, room_id: str, comment_id: str, emoji: str) -> None: self._client.post(f"{self._base}/rooms/{room_id}/comments/{comment_id}/reactions", json={"emoji": emoji})
    def remove_reaction(self, room_id: str, comment_id: str, reaction_id: str) -> None: self._client.delete(f"{self._base}/rooms/{room_id}/comments/{comment_id}/reactions/{reaction_id}")
    def list_annotations(self, room_id: str) -> list[Annotation]: return [Annotation(**a) for a in self._client.get(f"{self._base}/rooms/{room_id}/annotations")]
    def add_annotation(self, room_id: str, type: str, data: dict, **kwargs: Any) -> Annotation: return Annotation(**self._client.post(f"{self._base}/rooms/{room_id}/annotations", json={"type": type, "data": data, **kwargs}))
    def update_annotation(self, room_id: str, annotation_id: str, **kwargs: Any) -> Annotation: return Annotation(**self._client.patch(f"{self._base}/rooms/{room_id}/annotations/{annotation_id}", json=kwargs))
    def remove_annotation(self, room_id: str, annotation_id: str) -> None: self._client.delete(f"{self._base}/rooms/{room_id}/annotations/{annotation_id}")
    def clear_annotations(self, room_id: str) -> None: self._client.delete(f"{self._base}/rooms/{room_id}/annotations")
    def start_recording(self, room_id: str) -> dict: return self._client.post(f"{self._base}/rooms/{room_id}/recording/start")
    def stop_recording(self, room_id: str) -> dict: return self._client.post(f"{self._base}/rooms/{room_id}/recording/stop")
    def get_recording_status(self, room_id: str) -> dict: return self._client.get(f"{self._base}/rooms/{room_id}/recording")
